documentation/specs/task-isolation-and-dependencies.md

```diff
@@ -0,0 +1,38 @@
+# Task isolation
+## Problem definition
+Tasks in MSBuild are dynamically loaded assemblies with potentially separate and colliding dependency trees. Currently MSBuild on .NET Core has no isolation between tasks and as such only one version of any given assembly can be loaded. Prime example of this is Newtonsoft.Json which has multiple versions, but all the tasks must agree on it to work.
+This problem is also described in microsoft/msbuild#1754.
+
+## Possible solution
+Use [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.2) (ALC) to provide binding isolation for tasks. Each task would be loaded into its own ALC instance.
+* The ALC would resolve all dependencies of the tasks (see dependency resolution below)
+* ALC would fallback to the Default for dependencies which the task doesn't carry with itself (Framework and so on)
+* ALC would probably have to forcefully fallback for MSBuild assemblies since by default lot of tasks will carry these, but the system requires for the MSBuild assemblies to be shared.
+We would probably also want to load groups of tasks which belong together into the same ALC (for example based on their location on disk) to improve performance. This will need some care as there's no guarantee that two random tasks have compatible dependency trees.
+
+## Potential risks
+* Has some small probability of causing breaks. Currently all assemblies from all tasks are loaded into the default context and thus are "visible" to everybody. Tasks with following properties might not work:
+  * Task has a dependency on an assembly, but it doesn't declare this dependency in its .deps.json and this dependency gets loaded through some other task. This is mostly fixable by implementing probing similar to today's behavior.
+  * Two tasks from different assemblies which somehow rely on sharing certain types. If the new system decides to load these in isolation they won't share types anymore and might not work.
+* Performance - task isolation inherently (and by design) leads to loading certain assemblies multiple times. This increases memory pressure and causes additional JITing and other related work.
+
+## Additional consideration
+* None of these changes would have any effect on MSBuild on .NET Framework
+* Task isolation alone could be achieved on existing MSBuild
```

Comment:
With an ALC but no `.deps.json` handling?

Reply:
I think it would be nice if deps.json were optional to get ALC isolation. Can custom ALC behave as SetAppPaths option in runtimeconfig for default ALC / app?

Reply:
If deps.json files is missing, the resolver works with the contents of the published directory of the task.

Reply:
To answer @rainersigwald question: yes, we could add isolation logic into MSBuild even on .NET Core 2.*. MSBuild seems to have complex enough probing logic to make most tasks work (find their dependencies). The isolation would be about solving problems with tasks which have incompatible dependencies.
`.deps.json` handling makes the whole thing even more robust and enables additional scenarios, but it requires 3.0 APIs.